source/static_thread_pool.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+ #include <unifex/static_thread_pool.hpp>
+
+namespace unifex
+{
+
+static_thread_pool::static_thread_pool()
+: static_thread_pool(std::thread::hardware_concurrency())
+{}
+
+static_thread_pool::static_thread_pool(std::uint32_t threadCount) {
+    assert(threadCount > 0);
+
+    threads_.reserve(threadCount);
+    try {
+        threads_.emplace_back([this] { this->run(); });
+    } catch (...) {
+        close();
+    }
+}
+
+static_thread_pool::~static_thread_pool() {
+    close();
+}
+
+void static_thread_pool::run() noexcept {
+    std::unique_lock lk{mut_};
+    while (true) {
+        while (!queue_.empty()) {
+            auto* task = queue_.pop_front();
+            lk.unlock();
+            task->execute(task);
```

Comment:
exceptions? I would love to have a scoped reverse lock that would release the lock in the constructor and take it in the destructor. much safer to use than this manual locking

Reply:
The execute() function is noexcept so should be safe to manually unlock/lock here.
I agree it would be nice to have some RAII unlocker, though.